File: library/metaoutput.client/extension/AnyTool.cpp

#include "../All.hpp"

// extension::AnyTool ######################################################
void extension::AnyTool::Connect(MP_STRING application, MP_STRING extension)
{
    try
    {
        if (MP_STRING_EMPTY(application) == false)
        {
            MP_REGISTRY a_Context;
            {
                MP_REGISTRY_INITIALIZE(a_Context, MP_REGISTRY_ROOT_CURRENT_USER, "Software\\MetaPlatform\\APPLICATION\\", false);
            }
            if (MP_STRING_EMPTY(MP_REGISTRY_GET(a_Context, "USER.ID", "")))
            {
                MP_FILE_EXECUTE("https://www.metaoutput.net/requirements?application=" + application + "&extension=" + extension);
            }
            {
                MP_REGISTRY_FINALIZE(a_Context);
            }
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
}

void extension::AnyTool::Disconnect()
{
    try
    {
        if ((s_Items != nullptr) && (MP_VECTOR_EMPTY(s_Items) == false))
        {
            auto a_Size = MP_VECTOR_SIZE_GET(s_Items);
            for (auto i = 0; i < a_Size; i++)
            {
                auto a_Context = MP_VECTOR_GET(s_Items, 0);
                {
                    MP_VECTOR_DELETE(s_Items, 0);
                }
                {
                    Execute(a_Context->m_Name, pattern::data::Message::CONSTANT::PIPE::TERMINATE_REQUEST);
                }
            }
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
}

bool extension::AnyTool::Register(MP_STRING name, MP_PTR(AnyTool) handler)
{
    try
    {
        if (s_Items == nullptr)
        {
            s_Items = MP_NEW MP_VECTOR(AnyTool);
        }
        if ((s_Items != nullptr) && (handler != nullptr) && (MP_STRING_EMPTY(name) == false))
        {
            for (auto i = MP_VECTOR_SIZE_GET(s_Items) - 1; i >= 0; i--)
            {
                if (s_Items[i]->m_Name == name)
                {
                    return false;
                }
            }
            {
                handler->m_Name = name;
                handler->m_Thread = nullptr;
                handler->m_Handlers = MP_NEW MP_MAP(MP_PTR(atom::Trace::AnyHandler));
            }
            {
                MP_THREAD_INITIALIZE(handler->m_Thread, __ThreadExecute);
                MP_THREAD_NAME_SET(handler->m_Thread, name);
                MP_THREAD_APARTMENT_SET(handler->m_Thread, MP_THREAD_APARTMENT_STA);
                MP_THREAD_START(handler->m_Thread, handler);
            }
            {
                MP_VECTOR_APPEND(s_Items, handler);
            }
            return true;
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return false;
}

bool extension::AnyTool::Register(MP_STRING name, MP_STRING id, MP_PTR(atom::Trace::AnyHandler) handler)
{
    try
    {
        if ((s_Items != nullptr) && (handler != nullptr) && (MP_STRING_EMPTY(name) == false) && (MP_STRING_EMPTY(id) == false))
        {
            for (auto i = MP_VECTOR_SIZE_GET(s_Items) - 1; i >= 0; i--)
            {
                if (s_Items[i]->m_Name == name)
                {
                    if (MP_MAP_CONTAINS(s_Items[i]->m_Handlers, id) == false)
                    {
                        MP_MAP_APPEND(s_Items[i]->m_Handlers, id, handler);
                    }
                    return true;
                }
            }
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return false;
}

bool extension::AnyTool::Unregister(MP_STRING name)
{
    try
    {
        if ((s_Items != nullptr) && (MP_STRING_EMPTY(name) == false))
        {
            for (auto i = MP_VECTOR_SIZE_GET(s_Items) - 1; i >= 0; i--)
            {
                if (s_Items[i]->m_Name == name)
                {
                    MP_VECTOR_DELETE(s_Items, i);
                    return true;
                }
            }
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return false;
}

bool extension::AnyTool::Execute(MP_STRING name, MP_STRING value)
{
    try
    {
        if ((MP_STRING_EMPTY(value) == false) && (MP_STRING_EMPTY(MP_STRING_TRIM(value)) == false))
        {
            auto a_Context1 = (MP_PTR(MP_PIPE_CLIENT))nullptr;
            auto a_Context2 = (MP_PTR(MP_PIPE_WRITESTREAM))nullptr;
            {
                MP_PIPE_CLIENT_INITIALIZE(a_Context1, GetPipeName(name), MP_PIPE_DIRECTION_OUT);
                MP_PIPE_CLIENT_CONNECT(a_Context1, pattern::data::Message::CONSTANT::PIPE::TIMEOUT);
                MP_PIPE_WRITESTREAM_INITIALIZE(a_Context2, a_Context1);
            }
            if (MP_PIPE_CLIENT_CONNECTED(a_Context1))
            {
                MP_PIPE_WRITESTREAM_WRITE(a_Context2, value);
            }
            {
                MP_PIPE_WRITESTREAM_FINALIZE(a_Context2);
                MP_PIPE_CLIENT_FINALIZE(a_Context1);
            }
            return true;
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return false;
}

// Public ##############
MP_STRING extension::AnyTool::GetAttribute(MP_STRING attributes, MP_STRING name)
{
    try
    {
        auto a_Context = attributes;
        {
            auto a_Index = MP_STRING_INDEXOF(a_Context, pattern::data::Message::TML::COMMENT);
            if (a_Index >= 0)
            {
                a_Context = MP_STRING_SUBSTRING(a_Context, 0, a_Index);
            }
        }
        if (MP_STRING_CONTAINS(a_Context, "\t"))
        {
            a_Context = MP_STRING_REPLACE(a_Context, "\t", " ");
        }
        if (MP_STRING_CONTAINS(a_Context, "\r"))
        {
            a_Context = MP_STRING_REPLACE(a_Context, "\r", " ");
        }
        if (MP_STRING_CONTAINS(a_Context, "\n"))
        {
            a_Context = MP_STRING_REPLACE(a_Context, "\n", " ");
        }
        {
            auto a_Index = MP_STRING_INDEXOF(MP_STRING_UPPER(a_Context), pattern::data::Message::TML::ATTRIBUTE + MP_STRING_UPPER(name) + " ");
            if (a_Index >= 0)
            {
                auto a_Result = (MP_STRING)"";
                auto a_Size = MP_STRING_SIZE_GET(pattern::data::Message::TML::ATTRIBUTE + name);
                auto a_Index1 = MP_STRING_INDEXFROM(MP_STRING_UPPER(a_Context), pattern::data::Message::TML::ATTRIBUTE, a_Index + a_Size);
                if (a_Index1 < 0)
                {
                    a_Result = MP_STRING_TRIM(MP_STRING_SUBSTRING(a_Context, a_Index + a_Size, MP_STRING_SIZE_GET(a_Context) - a_Index - a_Size));
                }
                else
                {
                    a_Result = MP_STRING_TRIM(MP_STRING_SUBSTRING(a_Context, a_Index + a_Size, a_Index1 - a_Index - a_Size));
                }
                if (MP_STRING_CONTAINS(a_Result, pattern::data::Message::TML::NEW_LINE))
                {
                    a_Result = MP_STRING_REPLACE(a_Result, pattern::data::Message::TML::NEW_LINE, "\r\n");
                }
                return a_Result;
            }
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return "";
}

MP_STRING extension::AnyTool::GetPipeName(MP_STRING name)
{
    auto a_Context = ":" + MP_CONVERT_STRING_FROM_INT(MP_PROCESS_ID_GET(MP_PROCESS_CURRENT_GET()), 0);
    if (MP_STRING_CONTAINS(name, a_Context))
    {
        return "urn:metaoutput:tool:" + name;
    }
    else
    {
        return "urn:metaoutput:tool:" + name + a_Context;
    }
}

// Protected ###########
void extension::AnyTool::_Execute(MP_PTR(atom::Trace), MP_INT, MP_STRING)
{
}

// Private #############
void extension::AnyTool::__Execute(MP_PTR(atom::Trace) trace, MP_INT level, MP_STRING attributes)
{
    if (this != nullptr)
    {
        auto a_Id = GetAttribute(attributes, NAME::ATTRIBUTE::ID);
        if (MP_STRING_EMPTY(a_Id))
        {
            _Execute(trace, 0, m_Name);
            return;
        }
        if (MP_MAP_CONTAINS(m_Handlers, a_Id))
        {
            MP_MAP_GET(m_Handlers, a_Id)->Execute(trace, level, m_Name, a_Id, GetAttribute(attributes, NAME::ATTRIBUTE::ACTION), attributes);
        }
    }
}

// Private #############
void extension::AnyTool::MP_THREAD_CALLBACK_MAIN(__ThreadExecute, sender)
{
    auto a_Context = dynamic_cast<MP_PTR(AnyTool)>(sender);
    auto a_Trace = MP_NEW atom::Trace();
    auto a_IsFound = a_Context == nullptr;
    while (a_IsFound == false)
    {
        try
        {
            auto a_Context1 = (MP_PTR(MP_PIPE_SERVER))nullptr;
            auto a_Context2 = (MP_PTR(MP_PIPE_READSTREAM))nullptr;
            {
                MP_PIPE_SERVER_INITIALIZE(a_Context1, extension::AnyTool::GetPipeName(a_Context->m_Name), MP_PIPE_DIRECTION_IN, 128);
                MP_PIPE_READSTREAM_INITIALIZE(a_Context2, a_Context1);
            }
            while (a_IsFound == false)
            {
                try
                {
                    {
                        MP_PIPE_SERVER_CONNECT(a_Context1);
                    }
                    if (MP_PIPE_SERVER_CONNECTED(a_Context1))
                    {
                        auto a_Context3 = MP_PIPE_READSTREAM_READ(a_Context2);
                        {
                            MP_PIPE_SERVER_DISCONNECT(a_Context1);
                        }
                        if (a_Context3 == pattern::data::Message::CONSTANT::PIPE::TERMINATE_REQUEST)
                        {
                            a_IsFound = true;
                            break;
                        }
                        else
                        {
                            a_Context->__Execute(a_Trace->Clear(), 0, a_Context3);
                        }
                        {
                            a_Trace->CommitTransaction();
                            a_Trace->CancelTransaction();
                        }
                    }
                }
                catch (MP_PTR(MP_EXCEPTION) ex)
                {
                    a_Trace->CommitTransaction();
                    a_Trace->CancelTransaction();
                    a_Trace->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
                    MP_THREAD_SLEEP(pattern::data::Message::CONSTANT::PIPE::TIMEOUT);
                    break;
                }
            }
            {
                MP_PIPE_READSTREAM_FINALIZE(a_Context2);
                MP_PIPE_SERVER_FINALIZE(a_Context1);
            }
        }
        catch (MP_PTR(MP_EXCEPTION) ex)
        {
            a_Trace->CommitTransaction();
            a_Trace->CancelTransaction();
            a_Trace->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
            break;
        }
    }
}

File: library/metaoutput.client/extension/AnyTool.cs
#pragma warning disable CA1707 // Identifiers should not contain underscores

using atom;
using Microsoft.Win32;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.IO.Pipes;
using System.Threading;

namespace extension
{
#if RELEASE
    [System.Diagnostics.DebuggerStepThrough]
#endif
    public class AnyTool : basic.AnyObject
    {
        private string m_Name;
        private Thread m_Thread;
        private Dictionary<string, Trace.AnyHandler> m_Handlers;
        private static List<AnyTool> s_Items;

        public new class NAME : basic.AnyObject.NAME
        {
            public class ACTION
            {
                public const string CHANGE = "CHANGE";
                public const string CLICK = "CLICK";
                public const string ENTER = "ENTER";
                public const string LEAVE = "LEAVE";
                public const string MOVE = "MOVE";
                public const string WHEEL = "WHEEL";
            }

            public class ATTRIBUTE
            {
                public const string ACTION = "ACTION";
                public const string CURSOR_X = "CURSOR.X";
                public const string CURSOR_Y = "CURSOR.Y";
                public const string ID = "ID";
                public const string SELECTION_X = "SELECTION.X";
                public const string SELECTION_Y = "SELECTION.Y";
                public const string STATE = "STATE";
                public const string VALUE = "VALUE";
            }
        }

        public new class STATE : AnyPreview.STATE
        {
        }

        public static void Connect(string application, string extension)
        {
            try
            {
                if (!string.IsNullOrEmpty(application))
                {
                    var a_Context = Registry.CurrentUser.CreateSubKey("Software\\MetaPlatform\\APPLICATION\\");
                    if (string.IsNullOrEmpty(a_Context.GetValue("USER.ID", "").ToString()))
                    {
                        System.Diagnostics.Process.Start("https://www.metaoutput.net/requirements?application=" + application + "&extension=" + extension);
                    }
                    {
                        a_Context.Close();
                    }
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
        }

        public static void Disconnect()
        {
            try
            {
                if (s_Items != null)
                {
                    for (var i = s_Items.Count - 1; i >= 0; i--)
                    {
                        var a_Context = s_Items[i];
                        {
                            s_Items.RemoveAt(i);
                        }
                        {
                            Execute(a_Context.m_Name, CONSTANT.MESSAGE.PIPE.TERMINATE_REQUEST);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
        }

        public static bool Register(string name, string id, Trace.AnyHandler handler)
        {
            try
            {
                if (s_Items != null && handler != null && !string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(id))
                {
                    for (var i = s_Items.Count - 1; i >= 0; i--)
                    {
                        if (s_Items[i].m_Name == name)
                        {
                            if (!s_Items[i].m_Handlers.ContainsKey(id))
                            {
                                s_Items[i].m_Handlers.Add(id, handler);
                            }
                            return true;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return false;
        }

        public static bool Register(string name, AnyTool handler)
        {
            try
            {
                if (s_Items == null)
                {
                    s_Items = new List<AnyTool>();
                }
                if (handler != null && !string.IsNullOrEmpty(name))
                {
                    for (var i = s_Items.Count - 1; i >= 0; i--)
                    {
                        if (s_Items[i].m_Name == name)
                        {
                            return false;
                        }
                    }
                    {
                        handler.m_Name = name;
                        handler.m_Handlers = new Dictionary<string, Trace.AnyHandler>();
                        handler.m_Thread = new Thread(__ThreadExecute);
                        handler.m_Thread.Name = name;
                        handler.m_Thread.TrySetApartmentState(ApartmentState.STA);
                        handler.m_Thread.Start(handler);
                    }
                    {
                        s_Items.Add(handler);
                    }
                    return true;
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return false;
        }

        public static bool Unregister(string name)
        {
            try
            {
                if (s_Items != null && !string.IsNullOrEmpty(name))
                {
                    for (var i = s_Items.Count - 1; i >= 0; i--)
                    {
                        if (s_Items[i].m_Name == name)
                        {
                            s_Items.RemoveAt(i);
                            return true;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return false;
        }

        public static bool Execute(string name, string value)
        {
            try
            {
                if (!string.IsNullOrEmpty(value) && !string.IsNullOrEmpty(value.Trim()))
                {
                    var a_Context = new NamedPipeClientStream(".", GetPipeName(name), PipeDirection.Out);
                    {
                        a_Context.Connect(CONSTANT.MESSAGE.PIPE.TIMEOUT);
                    }
                    {
                        var a_Context1 = new StreamWriter(a_Context);
                        if (a_Context.IsConnected)
                        {
                            a_Context1.Write(value);
                            a_Context1.Flush();
                        }
                        {
                            a_Context1.Close();
                        }
                    }
                    {
                        a_Context.Close();
                    }
                    return true;
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return false;
        }

        public static string GetAttribute(string attributes, string name)
        {
            try
            {
                var a_Context = attributes;
                {
                    var a_Index = a_Context.IndexOf(CONSTANT.MESSAGE.TML.COMMENT, StringComparison.InvariantCulture);
                    if (a_Index >= 0)
                    {
                        a_Context = a_Context.Substring(0, a_Index);
                    }
                    if (a_Context.Contains("\t"))
                    {
                        a_Context = a_Context.Replace("\t", " ");
                    }
                    if (a_Context.Contains("\r"))
                    {
                        a_Context = a_Context.Replace("\r", " ");
                    }
                    if (a_Context.Contains("\n"))
                    {
                        a_Context = a_Context.Replace("\n", " ");
                    }
                }
                {
                    var a_Index = a_Context.IndexOf(CONSTANT.MESSAGE.TML.ATTRIBUTE + name.ToUpperInvariant() + " ", StringComparison.CurrentCultureIgnoreCase);
                    if (a_Index >= 0)
                    {
                        var a_Size = (CONSTANT.MESSAGE.TML.ATTRIBUTE + name).Length;
                        var a_Index1 = a_Context.IndexOf(CONSTANT.MESSAGE.TML.ATTRIBUTE, a_Index + a_Size, StringComparison.CurrentCultureIgnoreCase);
                        var a_Result = ((a_Index1 >= 0) ? a_Context.Substring(a_Index + a_Size, a_Index1 - a_Index - a_Size).Trim() : a_Context.Substring(a_Index + a_Size, a_Context.Length - a_Index - a_Size).Trim());
                        if (a_Result.Contains(CONSTANT.MESSAGE.TML.NEW_LINE))
                        {
                            a_Result = a_Result.Replace(CONSTANT.MESSAGE.TML.NEW_LINE, "\r\n");
                        }
                        return a_Result;
                    }
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return "";
        }

        public static string GetPipeName(string value)
        {
            var a_Context = ":" + System.Diagnostics.Process.GetCurrentProcess().Id.ToString("d0", CultureInfo.InvariantCulture);
            if (value.Contains(a_Context))
            {
                return "urn:metaoutput:tool:" + value;
            }
            return "urn:metaoutput:tool:" + value + a_Context;
        }

        protected virtual void _Execute(Trace trace, int level, string name)
        {
        }

        private void __Execute(Trace trace, int level, string value)
        {
            var a_Context = GetAttribute(value, NAME.ATTRIBUTE.ID);
            if (string.IsNullOrEmpty(a_Context))
            {
                _Execute(trace, 0, m_Name);
                return;
            }
            if (m_Handlers.TryGetValue(a_Context, out Trace.AnyHandler a_Result))
            {
                a_Result.Execute(trace, level, m_Name, a_Context, GetAttribute(value, NAME.ATTRIBUTE.ACTION), value);
            }
        }

        private static void __ThreadExecute(object sender)
        {
            var a_Context = sender as AnyTool;
            var a_Trace = new Trace();
            for (var a_IsFound = a_Context == null; !a_IsFound;)
            {
                try
                {
                    var a_Context1 = new NamedPipeServerStream(GetPipeName(a_Context.m_Name), PipeDirection.In, 128);
                    var a_Context2 = new StreamReader(a_Context1);
                    while (!a_IsFound)
                    {
                        try
                        {
                            {
                                a_Context1.WaitForConnection();
                            }
                            if (a_Context1.IsConnected)
                            {
                                var a_Context3 = a_Context2.ReadToEnd();
                                {
                                    a_Context1.Disconnect();
                                }
                                if (a_Context3 == CONSTANT.MESSAGE.PIPE.TERMINATE_REQUEST)
                                {
                                    a_IsFound = true;
                                    break;
                                }
                                else
                                {
                                    a_Trace.CancelTransaction();
                                    a_Context.__Execute(a_Trace, 0, a_Context3);
                                    a_Trace.CommitTransaction();
                                }
                            }
                        }
                        catch (Exception ex)
                        {
                            a_Trace.CommitTransaction();
                            a_Trace.SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
                            Thread.Sleep(CONSTANT.MESSAGE.PIPE.TIMEOUT);
                            break;
                        }
                    }
                    {
                        a_Context2.Close();
                        a_Context1.Close();
                    }
                }
                catch (Exception ex)
                {
                    a_Trace.CommitTransaction();
                    a_Trace.SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
                    break;
                }
            }
        }
    }
}
